src/app/users/login.ts

```diff
@@ -0,0 +1,126 @@
+import {
+  getFailedLoginAttemptPerIpLimits,
+  getFailedLoginAttemptPerPhoneLimits,
+} from "@config/app"
+import { CouldNotFindUserFromPhoneError, UnknownRepositoryError } from "@domain/errors"
+import { RateLimitPrefix } from "@domain/rate-limit"
+import {
+  RateLimiterExceededError,
+  UserLoginIpRateLimiterExceededError,
+  UserLoginPhoneRateLimiterExceededError,
+} from "@domain/rate-limit/errors"
+import { createToken } from "@services/jwt"
+import { UsersRepository } from "@services/mongoose"
+import { PhoneCodesRepository } from "@services/mongoose/phone-code"
+import { TwilioClient } from "@services/twilio"
+import { RedisRateLimitService } from "@services/rate-limit"
+import { isTestAccountPhoneAndCode } from "."
+
+export const login = async ({
+  phone,
+  code,
+  logger,
+  ip,
+}: {
+  phone: PhoneNumber
+  code: PhoneCode
+  logger: Logger
+  ip: IpAddress
+}): Promise<JwtToken | ApplicationError> => {
+  const subLogger = logger.child({ topic: "login" })
+
+  {
+    const limitOk = await checkFailedLoginAttemptPerIpLimits(ip)
+    if (limitOk instanceof Error) return limitOk
+  }
+
+  {
+    const limitOk = await checkFailedLoginAttemptPerPhoneLimits(phone)
+    if (limitOk instanceof Error) return limitOk
+  }
+
+  // TODO:
+  // add fibonachi on failed login
+  // https://github.com/animir/node-rate-limiter-flexible/wiki/Overall-example#dynamic-block-duration
+
+  const validCode = await isCodeValid({ phone, code })
+  if (validCode instanceof Error) return validCode
+
+  await rewardFailedLoginAttemptPerIpLimits(ip)
+
+  const userRepo = UsersRepository()
+  let user: RepositoryError | User = await userRepo.findByPhone(phone)
+
+  if (user instanceof CouldNotFindUserFromPhoneError) {
+    subLogger.info({ phone }, "new user signup")
+
+    const userRaw = { phone } as NewUserInfo
```

Comment:
can you explain an example on how this could be dangerous?

Reply:
It is dangerous because it exposes the possibility for runtime errors where they could be caught at compile time if you don't use the cast.

EG: say the `NewUserInfo` type evolves and we want to add a fiel:
```
type NewUserInfo
{
  phone: string
  anotherField: number
}
```

now you pass the new info to the repository:
```
const newUser = { phone } as NewUserInfo
repo.persistNew(newUser)
```
And the repo function is defined as:
```
persistNew = ({phone, anotherField} :NewUserInfo) => {
 anotherField += 5
}
```

This is just a simple example but its highlighting that because of the cast this will compile - and not only that it will ALWAYS compile even if the definition of `NewUserInfo` changes.

Reply:
ok makes sense

Reply:
fixed